### ISSUE TITLE
[JENKINS-54732] Revision information produced by pipline 'checkout' operation isn't parallel safe

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-plugin#3363]
<p>I recently ran some test builds on our build farm to check out the "new" return data produced by the scm "checkout" function. In our case we are using Git as our source repository and are doing checkouts using code that looks like this:</p>

<p> </p>

<p>def branchName = "MyBranch"<br/>
def res = checkout (<br/>
    changelog: true,<br/>
    poll: false,<br/>
    scm: [<br/>
        $class: 'GitSCM',<br/>
        branches: [<span class="error">&#91;name: branchName&#93;</span>],<br/>
        browser: <span class="error">&#91;$class: &#39;GitLab&#39;, repoUrl: "path/to/repo&#39;, version: "10.0"&#93;</span>,<br/>
        doGenerateSubmoduleConfigurations: false,<br/>
        extensions: [<br/>
            <span class="error">&#91;$class: &#39;CloneOption&#39;, depth: 0, noTags: false, reference: &#39;&#39;, shallow: false, timeout: 90&#93;</span>,<br/>
        ],<br/>
        submoduleCfg: [],<br/>
 <br/>
        userRemoteConfigs: [<span class="error">&#91;credentialsId: &#39;git_creds&#39;, url: "path/to/repo.git"&#93;</span>]<br/>
]<br/>
)<br/>
echo "Results from checkout are " + res.toString()<br/>
 <br/>
This code works well enough on it's own, until you try to run it in parallel. In our case we have 2 parallel stages that are set up to check out code from the same Git repository but that use 2 different branches. When you examine the output produced by the "echo" statement at the end of the snippet you will see that the revision information returned (ie: res.GIT_COMMIT) reflects the correct revision number for the "first" clone operation that runs, but it does NOT for the "second" clone operation (dependent upon which of the parallel stages runs first). The second clone operation to run seems to be returning result data from the first clone operation, almost like the two clone processes are sharing common data behind the scenes or something. This makes it impossible to rely on the results of a "checkout" call when performing the operation in parallel.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=leedega">leedega</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54732" target="_blank">Revision information produced by pipline 'checkout' operation isn't parallel safe</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: git-plugin, pipeline
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 1
<li><b>watchers</b>: 4
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I recently ran some test builds on our build farm to check out the "new" return data produced by the scm "checkout" function. In our case we are using Git as our source repository and are doing checkouts using code that looks like this:</p>

<p> </p>

<p>def branchName = "MyBranch"
def res = checkout (
    changelog: true,
    poll: false,
    scm: [
        $class: 'GitSCM',
        branches: [<span class="error">&#91;name: branchName&#93;</span>],
        browser: <span class="error">&#91;$class: &#39;GitLab&#39;, repoUrl: &quot;path/to/repo&#39;, version: &quot;10.0&quot;&#93;</span>,
        doGenerateSubmoduleConfigurations: false,
        extensions: [
            <span class="error">&#91;$class: &#39;CloneOption&#39;, depth: 0, noTags: false, reference: &#39;&#39;, shallow: false, timeout: 90&#93;</span>,
        ],
        submoduleCfg: [],
 
        userRemoteConfigs: [<span class="error">&#91;credentialsId: &#39;git_creds&#39;, url: &quot;path/to/repo.git&quot;&#93;</span>]
]
)
echo "Results from checkout are " + res.toString()
 
This code works well enough on it's own, until you try to run it in parallel. In our case we have 2 parallel stages that are set up to check out code from the same Git repository but that use 2 different branches. When you examine the output produced by the "echo" statement at the end of the snippet you will see that the revision information returned (ie: res.GIT_COMMIT) reflects the correct revision number for the "first" clone operation that runs, but it does NOT for the "second" clone operation (dependent upon which of the parallel stages runs first). The second clone operation to run seems to be returning result data from the first clone operation, almost like the two clone processes are sharing common data behind the scenes or something. This makes it impossible to rely on the results of a "checkout" call when performing the operation in parallel.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins v2.148<br/>
Pipeline Plugin v2.5<br/>
Git Plugin v3.9.1<br/>
Git Client Plugin v2.7.3
```
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>leedega:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54732?focusedId=354285&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-354285">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I am linking defect <a href="https://issues.jenkins.io/browse/JENKINS-39968" title="git checkouts are not pipeline-parallel safe" class="issue-link" data-issue-key="JENKINS-39968">JENKINS-39968</a> to this one since I have a gut feeling that the root cause of my defect here is probably related to the root cause of this other defect as well. Assuming there is some sort of shared-state within the "checkout" function, and this shared-state is not thread safe, then it could explain why someone might get inconsistent results elsewhere - like in the change logs - when running checkouts in parallel.</p>
</pre>
</details>

</ul>
</details>

<p>I am linking defect <a href="https://issues.jenkins.io/browse/JENKINS-39968" title="git checkouts are not pipeline-parallel safe" class="issue-link" data-issue-key="JENKINS-39968">JENKINS-39968</a> to this one since I have a gut feeling that the root cause of my defect here is probably related to the root cause of this other defect as well. Assuming there is some sort of shared-state within the "checkout" function, and this shared-state is not thread safe, then it could explain why someone might get inconsistent results elsewhere - like in the change logs - when running checkouts in parallel.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>leedega:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54732?focusedId=354286&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-354286">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Linking another related task to this one, which I think may have been closed solely due to the lack of detail on the bug report. However, based on the brief description of the problem it sounds like it's the same thing I've encountered here. That means I'm likely not the first person to report this problem.</p>
</pre>
</details>

</ul>
</details>

<p>Linking another related task to this one, which I think may have been closed solely due to the lack of detail on the bug report. However, based on the brief description of the problem it sounds like it's the same thing I've encountered here. That means I'm likely not the first person to report this problem.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54732?focusedId=354287&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-354287">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>In  the short term, you'll need to use shell commands to extract the information you need rather than relying on the return values from `checkout scm` .  That should allow you to get the information you need without waiting for a fix.</p>
</pre>
</details>

</ul>
</details>

<p>In  the short term, you'll need to use shell commands to extract the information you need rather than relying on the return values from `checkout scm` .  That should allow you to get the information you need without waiting for a fix.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>leedega:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54732?focusedId=354290&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-354290">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Actually it's funny you mention that. This is what we are currently doing and have been doing for quite some time, since before the feature was added to the "checkout" function to return the metadata. I just recently decided to try out using the return data to simplify some of our scripts and so forth and wanted to leverage the return data produced by this build step.</p>

<p>The fact that the step produces return data that is incorrect, at least in the case of performing operations in parallel, suggests to me the handling of the data is fundamentally broken in some way. That being the case people shouldn't be relying on it being correct. Even if they rely on it when running serialized operations, they are likely to hit odd, hard to debug problems if they try to extend their build scripts down the road to parallelize them. It's just a ticking time bomb waiting to happen. That being the case I would personally suggest removing the return data completely from to avoid confusion and error until the functionality can be implemented correctly.</p>

<p>Also, if I am correct and the underlying root cause of this problem is in fact related to the other problems I've linked to above then that also suggests that the root cause of this problem - whatever that may be - may affect systems other than just the return data produced by the build task. That being the case the importance / criticality of this defect really should be reconsidered since it may affect even more areas, like changeset reporting ... and possibly others which may be even more severe (ie: commit triggers, etc. etc.).</p>

<p>If an assessment has already been completed and the root cause of this defect is known to be of minimal impact and limited to just the metadata being returned, then I agree - using shell scripts to gather checkout information is a viable workaround. Perhaps you could provide a bit more context surrounding the problem to set users' minds at ease in this regard.</p>
</pre>
</details>

</ul>
</details>

<p>Actually it's funny you mention that. This is what we are currently doing and have been doing for quite some time, since before the feature was added to the "checkout" function to return the metadata. I just recently decided to try out using the return data to simplify some of our scripts and so forth and wanted to leverage the return data produced by this build step.</p>

<p>The fact that the step produces return data that is incorrect, at least in the case of performing operations in parallel, suggests to me the handling of the data is fundamentally broken in some way. That being the case people shouldn't be relying on it being correct. Even if they rely on it when running serialized operations, they are likely to hit odd, hard to debug problems if they try to extend their build scripts down the road to parallelize them. It's just a ticking time bomb waiting to happen. That being the case I would personally suggest removing the return data completely from to avoid confusion and error until the functionality can be implemented correctly.</p>

<p>Also, if I am correct and the underlying root cause of this problem is in fact related to the other problems I've linked to above then that also suggests that the root cause of this problem - whatever that may be - may affect systems other than just the return data produced by the build task. That being the case the importance / criticality of this defect really should be reconsidered since it may affect even more areas, like changeset reporting ... and possibly others which may be even more severe (ie: commit triggers, etc. etc.).</p>

<p>If an assessment has already been completed and the root cause of this defect is known to be of minimal impact and limited to just the metadata being returned, then I agree - using shell scripts to gather checkout information is a viable workaround. Perhaps you could provide a bit more context surrounding the problem to set users' minds at ease in this regard.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54732?focusedId=354291&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-354291">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Unfortunately, I don't have more context surrounding the problem.  I haven't investigated it further than is described in the bug reports that you linked.</p>
</pre>
</details>

</ul>
</details>

<p>Unfortunately, I don't have more context surrounding the problem.  I haven't investigated it further than is described in the bug reports that you linked.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54732?focusedId=358471&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-358471">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Looks more likely to be a duplicate of <a href="https://issues.jenkins.io/browse/JENKINS-53346" title="GitSCM checkout returns same values on second different call" class="issue-link" data-issue-key="JENKINS-53346"><del>JENKINS-53346</del></a>.</p>
</pre>
</details>

</ul>
</details>

<p>Looks more likely to be a duplicate of <a href="https://issues.jenkins.io/browse/JENKINS-53346" title="GitSCM checkout returns same values on second different call" class="issue-link" data-issue-key="JENKINS-53346"><del>JENKINS-53346</del></a>.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/git-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-39968%22">JENKINS-39968</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/git-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-29193%22">JENKINS-29193</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/git-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-53346%22">JENKINS-53346</a>]</i>
